lib/src/css_mirroring/util_functions.dart

```diff
@@ -0,0 +1,109 @@
+bool _isRuleRemovable(_PendingRemovals removals, RuleSet rule) =>
+    removals._removalStartEndLocations.length ==
+        rule.declarationGroup.declarations.length;
+
+/// Checks if the declaration has to be removed based on the the Retention mode.
+bool _shouldRemoveDecl(
+    _RetentionMode mode, Declaration original, Declaration flipped) {
+  final bool isEqual = _areDeclarationsTextEqual(original, flipped);
+  return mode == _RetentionMode.keepBidiNeutral ? !isEqual : isEqual;
+}
+
+
+int _getNodeStart(TreeNode node) {
+  if (node is RuleSet) return node.span.start.offset;
+// In case of Directives since the node span start does not include '@'
+// so additional -1 is required.
+  return node.span.start.offset - 1;
+}
+
+int _getRuleSetEnd(List<RuleSet> ruleSets, int iTopLevel, int parentEnd) {
+  final int end = iTopLevel < ruleSets.length - 1
+      ? _getNodeStart(ruleSets[iTopLevel + 1])
+      : parentEnd;
+  return end;
+}
+
+int getDeclarationEnd(String source, int iDecl, List<Declaration> decls) {
+  if (iDecl < decls.length - 1) {
+    return decls[iDecl + 1].span.start.offset;
+  }
+
+  final int fileLength = source.length;
+  int fromIndex = decls[iDecl].span.end.offset;
+  try {
+    while (fromIndex + 1 < fileLength) {
+      if (source.substring(fromIndex, fromIndex + 2) == '/*') {
+        while (source.substring(fromIndex, fromIndex + 2) != '*/') {
+          fromIndex++;
+        }
+      } else if (source[fromIndex] == '}') {
+        return fromIndex;
+      }
+      fromIndex++;
+    }
+  } catch (_) {
+    throw new ArgumentError('Invalid CSS');
+  }
+  // Case when it doesnot find the end of declaration till file end.
+  if(source[fromIndex] == '}')
+    return fromIndex;
+  throw new ArgumentError('Declaration end not found');
+}
+
+List<TreeNode> _getUsedTopLevels(_RetentionMode mode,
+    TopLevelsPair _topLevelsPair) =>
+    (mode == _RetentionMode.keepFlippedBidiSpecific)
+        ? _topLevelsPair._flippedTopLevels
+        : _topLevelsPair._originalTopLevels;
+
+//_getUsedDirective(_RetentionMode mode, originalDirective, flippedDirective) =>
+//    (mode == _RetentionMode.keepFlippedBidiSpecific)
+//        ? flippedDirective
+//        : originalDirective;
+
+void _removeDirective(TextEditTransaction trans, TreeNode topLevel) =>
+  trans.edit(_getNodeStart(topLevel), topLevel.span.end.offset, '');
+
+void _prependDirectionToRuleSet(
+    EditConfiguration _editConfig, RuleSet _ruleSet) {
+  _editConfig.transaction.edit(_ruleSet.span.start.offset, _ruleSet.span.start.offset,
+      ':host-context([dir="${enumName(_editConfig.targetDirection)}"]) ');
+}
+
+/// Removes a rule from the transaction.
+void _removeRuleSet(
+    TextEditTransaction trans, List<RuleSet> rulesets, int iTopLevel) {
+  trans.edit(_getNodeStart(rulesets[iTopLevel]),
+      _getRuleSetEnd(rulesets, iTopLevel, trans.file.length), '');
+}
+
+List<Declaration> getDecls(Treenode a) => a.declarationGroup.declarations;
```

Comment:
TreeNode, not Treenode (you're probably getting an analyzer warning here, and checked mode considers this as dynamic since it's an unknown type)